src/python/pants/engine/internals/parametrize.py

```diff
@@ -0,0 +1,111 @@
+            if not isinstance(arg, str):
+                raise Exception(
+                    f"In {self}:\n  Positional arguments must be strings, but "
+                    f"`{arg}` was a `{type(arg).__name__}`."
+                )
+            previous_arg = parameters.get(arg)
+            if previous_arg is not None:
+                raise Exception(
+                    f"In {self}:\n  Positional arguments cannot have the same name as "
+                    f"keyword arguments. `{arg}` was also defined as `{arg}={previous_arg}`."
+                )
+            banned_chars = BANNED_CHARS_IN_PARAMETERS & set(arg)
+            if banned_chars:
+                raise Exception(
+                    f"In {self}:\n  Positional argument `{arg}` contained separator characters "
+                    f"(`{'`,`'.join(banned_chars)}`).\n\n"
+                    "To use `{arg}` as a parameter, you can pass it as a keyword argument to "
+                    "give it an alias. For example: `parametrize(short_memorable_name='{arg}')`"
+                )
+            parameters[arg] = arg
+        return parameters
+
+    @classmethod
+    def expand(
+        cls, address: Address, fields: dict[str, Any | Parametrize]
+    ) -> Iterator[tuple[Address, dict[str, Any]]]:
+        """Produces the cartesian product of fields for the given possibly-Parametrized fields.
+
+        Only one level of expansion is performed: if individual field values might also contain
+        Parametrize instances (in particular: an `overrides` field), expanding those will require
+        separate calls.
+        """
+        try:
+            parametrized: list[list[tuple[str, str, Any]]] = [
+                [
+                    (field_name, alias, field_value)
+                    for alias, field_value in v.to_parameters().items()
+                ]
+                for field_name, v in fields.items()
+                if isinstance(v, Parametrize)
+            ]
+            if not parametrized:
+                yield (address, fields)
+                return
+
+            non_parametrized = [
+                (field_name, field_value)
+                for field_name, field_value in fields.items()
+                if not isinstance(field_value, Parametrize)
+            ]
+
+            for parametrized_args in itertools.product(*parametrized):
+                expanded_address = address.parametrize(
+                    {field_name: alias for field_name, alias, _ in parametrized_args}
+                )
+                expanded_fields: dict[str, Any] = dict(
+                    (
+                        *non_parametrized,
+                        *(
+                            (field_name, field_value)
+                            for field_name, _, field_value in parametrized_args
+                        ),
+                    )
+                )
+                yield (expanded_address, expanded_fields)
+
+        except Exception as e:
```

Comment:
This is a really long try block, which makes it hard to reason about where the failures can come from. Is it possible to make this tighter?